### PR TITLE
Write output in last time step

### PR DESCRIPTION
**Pull request by mcwimm**
Small change to write output after last time step (in case all trees died).
Imho, it's a required feature to benchmark mortality concepts  effeciently.

Just realized that it doesn't work... @jbathmann do you have an idea how to implement this?

### CODE REVIEW

**Comment by jbathmann** [comment]
In line 90 of TreeDynamicTimeStepping, the timeloop is closed. Does it work, when you change it to

```
    def finish(self, time):
        self.visualization.show(time)
        tree_groups = self.population.getTreeGroups()

        self.tree_output.writeOutput(tree_groups, time, is_last_timestep=True)
```
?

I think you need to adapt the external timestepper to, right?

**Comment by mcwimm** [comment]
The first attemp is not working as we need the information from the previous time step (before all trees died). If we make a deep copy of `tree_groups` in each time step it doubles the running time.

*Test*
- SimpleBettina, 100 time steps, 100 trees
- Running time 
   + without copy: ~3 secs
   + with copy: ~7 secs

File: TimeLoopLib/TreeDynamicTimeStepping.py [hunk 1]
```diff
@@ -53,6 +55,7 @@ def step(self, t_start, t_end, update_ag, update_bg):
                 self.belowground_resources = (
                     self.belowground_competition.getBelowgroundResources())
         j = 0
+        number_of_trees = 0
```
[the remark this answers is not in the document]
Comment (jbathmann) [reply TimeLoopLib/TreeDynamicTimeStepping.py:58]: But can't you move line 80 up and remove it?

File: TimeLoopLib/TreeDynamicTimeStepping.py [hunk 2]
```diff
@@ -80,8 +83,16 @@ def step(self, t_start, t_end, update_ag, update_bg):
...
             print("INFO: MANGA execution stopped because all trees died and "
                   "no new tree were recruited.")
+
+            self.tree_output.writeOutput(self._previous_tree_groups,
+                                         t_start,
+                                         it_is_last_ts=True)
+
             exit()
         self.visualization.update(tree_groups, t_end)
+        self._previous_tree_groups = copy.deepcopy(tree_groups)
```
Comment (jbathmann) [inline TimeLoopLib/TreeDynamicTimeStepping.py:93]: I would define this earlier in the step (line 30). This way, it does not need to be saved as long and you can only define it locally. Does this add significant speedup?

**Comment by mcwimm** [comment]
Update: Thanks, @jbathmann !!
The new script writes an additional output when the last tree of a group dies.

File: TimeLoopLib/TreeDynamicTimeStepping.py [hunk 3]
```diff
@@ -71,6 +75,16 @@ def step(self, t_start, t_end, update_ag, update_bg):
                     kill_indices.append(i)
 
                 j += 1
+
+            # If all trees of a group died, make a copy of this tree set
+            if len(kill_indices) > 0:
+                if len(kill_indices) == tree_group.getNumberOfTrees():
+                    eliminated_tree_groups[tree_group.name] = copy.deepcopy(
```
Comment (mcwimm) [inline TimeLoopLib/TreeDynamicTimeStepping.py:82]: the script is doing what it is suppossed to do but I don't understand why we need the copy here.
Removing lines 82+83 and using 'tree_output' in line 84 produces an error as tree_output is empty when it's in the writing procedure.

File: TimeLoopLib/TreeDynamicTimeStepping.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mcwimm) [reply TimeLoopLib/TreeDynamicTimeStepping.py:58]: No, then we have still twice the amount of trees... or I don't get it...

[the remark this answers is not in the document]
Comment (mcwimm) [reply TimeLoopLib/TreeDynamicTimeStepping.py:58]: > In line 90 of TreeDynamicTimeStepping, the timeloop is closed. Does it work, when you change it to
> 
> ```
>     def finish(self, time):
>         self.visualization.show(time)
>         tree_groups = self.population.getTreeGroups()
> 
>         self.tree_output.writeOutput(tree_groups, time, is_last_timestep=True)
> ```
> 
> ?
> 
> I think you need to adapt the external timestepper, too. right?

In the external timestepper we don't have a finish-method... however, I need to recover my old files to test it. Anyway, we can solve this issue (if it excists) in another PR.

[the remark this answers is not in the document]
Comment (jbathmann) [reply TimeLoopLib/TreeDynamicTimeStepping.py:58]: :+1:

File: TimeLoopLib/TreeDynamicTimeStepping.py (hunk 3 above)
Comment (jbathmann) [reply TimeLoopLib/TreeDynamicTimeStepping.py:82]: As discussed directly:
- The script is not doing what it is supposed to do
- required to change into sth like 
```
if gruppozid:
    dying_group = ["dying"+group_name,tree_group]
    self.tree_output.writeOutput(dying_group, t_start, is_last_ts`True)
```

File: TreeOutputLib/TreeOutput.py [hunk 4]
```diff
@@ -138,13 +138,15 @@ def checkRequiredKey(self, key, args):
     #  For each timestep a file is created throughout the simulation.
     #  This function is only able to work, if the output directory exists and
     #  is empty at the begin of the model run
-    def writeOutput(self, tree_groups, time):
+    def writeOutput(self, tree_groups, time, it_is_last_ts=False):
```
Comment (jbathmann) [inline TreeOutputLib/TreeOutput.py:141]: After some thoughts I would suggest a more descriptive name for the keyword.
```suggestion
    def writeOutput(self, tree_groups, time, force_output=False):
```

Comment (mcwimm) [reply TreeOutputLib/TreeOutput.py:141]: good idea

File: TreeOutputLib/TreeOutput.py [hunk 5]
```diff
@@ -138,13 +138,15 @@ def checkRequiredKey(self, key, args):
     #  For each timestep a file is created throughout the simulation.
     #  This function is only able to work, if the output directory exists and
     #  is empty at the begin of the model run
-    def writeOutput(self, tree_groups, time):
+    def writeOutput(self, tree_groups, time, force_outpu=False):
```
Comment (jbathmann) [inline TreeOutputLib/TreeOutput.py:141]:
```suggestion
    def writeOutput(self, tree_groups, time, force_output=False):
```

File: TreeOutputLib/TreeOutput.py [hunk 6]
```diff
@@ -138,13 +138,15 @@ def checkRequiredKey(self, key, args):
...
         if self.output_each_nth_timestep is not None:
             self._output_counter = (self._output_counter %
                                     self.output_each_nth_timestep)
             self._it_is_output_time = (self._output_counter == 0)
         if self.output_times is not None:
             self._it_is_output_time = (time in self.output_times)
+        if force_outpu:
```
Comment (jbathmann) [inline TreeOutputLib/TreeOutput.py:148]:
```suggestion
        if force_output:
```

File: TimeLoopLib/TreeDynamicTimeStepping.py [hunk 7]
```diff
@@ -87,8 +101,9 @@ def step(self, t_start, t_end, update_ag, update_bg):
     def finish(self, time):
         self.visualization.show(time)
         tree_groups = self.population.getTreeGroups()
-
-        self.tree_output.writeOutput(tree_groups, time)
+        # Write output in last time step, even if not defined in the project
+        # file
+        self.tree_output.writeOutput(tree_groups, time, force_outpu=True)
```
Comment (jbathmann) [inline TimeLoopLib/TreeDynamicTimeStepping.py:106]:
```suggestion
        self.tree_output.writeOutput(tree_groups, time, force_output=True)
```

File: TimeLoopLib/TreeDynamicTimeStepping.py [hunk 8]
```diff
@@ -71,6 +75,16 @@ def step(self, t_start, t_end, update_ag, update_bg):
...
                 j += 1
+
+            # If all trees of a group died, make a copy of this tree set
+            if len(kill_indices) > 0:
+                if len(kill_indices) == tree_group.getNumberOfTrees():
+                    eliminated_tree_groups[tree_group.name] = copy.deepcopy(
+                        tree_group)
+                    self.tree_output.writeOutput(eliminated_tree_groups,
+                                                 t_start,
+                                                 force_outpu=True)
```
Comment (jbathmann) [inline TimeLoopLib/TreeDynamicTimeStepping.py:86]:
```suggestion
                                                 force_output=True)
```

**Review by jbathmann — APPROVED** [review]
Looks good!
Only the tests seem to fail atm. 
Option 1: Update SmallSetup3 and SmallSetup4 (which will be deprecated soon)
Option 2: Wait for megre of #195 and update corresponding reference files thereafter

**Comment by jbathmann** [comment]
@mcwimm

**Comment by mcwimm** [comment]
> @mcwimm

I choose option 2 :)

**Pull request merged**